Update Azure.AI.OpenAI to 2.1.0 and add tests

manvkaur opened the pull request:
This pull request introduces several breaking changes, dependency updates, and enhancements to the Azure Functions OpenAI Extension and its Java library. Key updates include renaming properties and classes for better clarity, adding managed identity support, and updating dependencies. Additionally, a new unit test project has been added, and documentation has been updated to reflect these changes.

### Breaking Changes and Enhancements:
* Renamed properties in bindings (`model` to `chatModel` and `embeddingsModel`, `connectionName` to `searchConnectionName` and `storeConnectionName`) and renamed `ChatMessage` to `AssistantMessage`. [[1]](diffhunk://#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR14-R25) [[2]](diffhunk://#diff-3be278d94fdaec6fe272a2848c2b9e66abad87455232f24bab4b46b36107a958R10-R20) [[3]](diffhunk://#diff-e396ab40126d8d1d432450dc2747fa228a4e0bdc45ce2de29ee2d82b9a3bbee1L64-R64) [[4]](diffhunk://#diff-e396ab40126d8d1d432450dc2747fa228a4e0bdc45ce2de29ee2d82b9a3bbee1R73-R80) [[5]](diffhunk://#diff-74a0514219b82c288ba18c098c3339e620768539108b4d2703a0a170f816c9e6L11-R31) [[6]](diffhunk://#diff-4f1a3dbdccc02d1cf9f51240a73ed02e984b21397f6774a4241a42c5633db760L23-R27)
* Updated Java library annotations, Assistant and Semantic Search bindings to include new configuration options (e.g., `temperature`, `topP`, `maxTokens`) and replaced `EmbeddingsOptions` with a `List<String>` for input in `EmbeddingsContext`. [[1]](diffhunk://#diff-e396ab40126d8d1d432450dc2747fa228a4e0bdc45ce2de29ee2d82b9a3bbee1R97-R131) [[2]](diffhunk://#diff-1b013df262e4e214484a48a0f555ceb80610b61f1f0936d0a430576dce74c21dL10-R23)


### New Features:
* Introduced `aiConnectionName` for managed identity support in bindings like `AssistantPost`, `EmbeddingsStore`, `SemanticSearch`and `TextCompletion`. [[1]](diffhunk://#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR14-R29) [[2]](diffhunk://#diff-e396ab40126d8d1d432450dc2747fa228a4e0bdc45ce2de29ee2d82b9a3bbee1R74-R81)
* Additional configuration options for AI service connections, including managed identity and fallback to environment variables.
* Added experimental `isReasoningModel` property to support reasoning models. [[1]](diffhunk://#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR14-R29) [[2]](diffhunk://#diff-e396ab40126d8d1d432450dc2747fa228a4e0bdc45ce2de29ee2d82b9a3bbee1R98-R139)


### Dependency Updates:
* Updated `Azure.AI.OpenAI` to version `2.1.0` and `azure-ai-openai` to version `1.0.0-beta.16`. [[1]](diffhunk://#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR14-R25) [[2]](diffhunk://#diff-399fef93b59e80488ed9e281b539e65fa3066f7c8ced1fb2e4c1000c3f5d5af2L91-R91)

### Testing and CI Enhancements:
* Added a new unit test project `WebJobsOpenAIUnitTests` to the solution and included it in the CI pipeline for build and test coverage. [[1]](diffhunk://#diff-a3a1680cfea45531d7e8bd91e764e4f4b814d645f897da4fe9f24a7cff3caab0R121-R122) [[2]](diffhunk://#diff-9358003ac8a7626af9f7a79802b49a32ef322427a270c08d736b358e2fa69965R23-R29)
* Add Samples Tests for:
  1. Embeddings
  5. Semantic Search
  6. Text Completion
* Add Unit Tests for:
  1. OpenAI Client Factory
  2. Assistant Service
  
### Documentation Updates:
* Updated `README.md` to reflect new requirements (e.g., `.NET 8 SDK`, Azurite) and detailed configuration options for AI service connections. Added examples for managed identity and configuration sections. [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L32-R100) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L273-R330)

### File and Class Renaming:
* Renamed `ChatMessage` to `AssistantMessage` in the Java library, Worker and WebJobs csproj, including corresponding updates to fields and methods. [[1]](diffhunk://#diff-74a0514219b82c288ba18c098c3339e620768539108b4d2703a0a170f816c9e6L11-R31) [[2]](diffhunk://#diff-74a0514219b82c288ba18c098c3339e620768539108b4d2703a0a170f816c9e6L82-L99) [[3]](diffhunk://#diff-4f1a3dbdccc02d1cf9f51240a73ed02e984b21397f6774a4241a42c5633db760L96-R96) [[4]](diffhunk://#diff-4f1a3dbdccc02d1cf9f51240a73ed02e984b21397f6774a4241a42c5633db760L159-R159)
* Removed reference to preview package Microsoft.Azure.WebJobs.Script.Abstractions and deleted unused BuiltInFunctionsProvider.cs

### Issues  Resolved:
  1. https://github.com/Azure/azure-functions-openai-extension/issues/104
  2. https://github.com/Azure/azure-functions-pyfx-planning/issues/395
  3. https://github.com/Azure/azure-functions-pyfx-planning/issues/164
  4. https://github.com/Azure/azure-functions-pyfx-planning/issues/269

### Performance impact:
1. Latency in first client creation as it has moved from start up to binding, later calls can retrieve from cache if exists.

Will add more test coverage in upcoming PRs.

Next PR:

- [ ] Python decorators
- [ ] Python samples
- [ ] Java samples (dependent on publish of java library part of current PR)

aishwaryabh commented on java-library/src/main/java/com/microsoft/azure/functions/openai/annotation/assistant/AssistantMessage.java:
```diff
@@ -8,38 +8,27 @@
  * <p>
- * Chat Message Entity which contains the content of the message, the role of the chat agent, and the name of the calling function if applicable.
+ * Chat Message Entity which contains the content of the message, the role of
+ * the chat agent, and the name of the calling function if applicable.
```
nit shouldn't be on a new line

aishwaryabh commented on src/Functions.Worker.Extensions.OpenAI/Assistants/ChatCompletionJsonConverter.cs:
```diff
@@ -1,23 +1,23 @@
 {
     static readonly ModelReaderWriterOptions modelReaderWriterOptions = new("J");
-    public override ChatCompletions Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+    public override ChatCompletion Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
```
nit: new line

aishwaryabh commented on src/WebJobs.Extensions.OpenAI/Assistants/AssistantService.cs:
```diff
@@ -171,110 +172,100 @@ public async Task<AssistantState> GetStateAsync(AssistantQueryAttribute assistan
+    }
+
+    // Helper methods
+    void ValidateAttributes(AssistantPostAttribute attribute)
```
helper methods should be internal

manvkaur replied:
these are more of private method as their only reference is within the same class.
for private method, access modifier specification is not required, by default if no access modifier is specified, method will be private.

should these still be update to internal access modifier?

aishwaryabh commented on src/WebJobs.Extensions.OpenAI/Embeddings/EmbeddingsConverter.cs:
```diff
@@ -47,11 +49,16 @@ async Task<EmbeddingsContext> ConvertCoreAsync(
-        this.logger.LogInformation("Sending OpenAI embeddings request: {request}", request.Input);
-        Response<OpenAISDK.Embeddings> response = await this.openAIClient.GetEmbeddingsAsync(request, cancellationToken);
-        this.logger.LogInformation("Received OpenAI embeddings count: {response}", response.Value.Data.Count);
+        List<string> input = await EmbeddingsHelper.BuildRequest(attribute.MaxOverlap,
```
On a quick glance, I think this code is duplicated in a few places. Is it possible to have a helper and just call that?

manvkaur replied:
Updated two occurrences, please check

aishwaryabh reviewed (approved):
Left a few nit comments but looks good overall!

TsuyoshiUshio commented:
Hi @manvkaur 

I'd like to understand the context why you changed the name as breaking change.

```
Renamed properties in bindings (model to chatModel and embeddingsModel, connectionName to searchConnectionName and storeConnectionName) and renamed ChatMessage to AssistantMessage. Updated managed identity support via the aiConnectionName property for Azure OpenAI
```

manvkaur commented:
> Hi @manvkaur
> 
> I'd like to understand the context why you changed the name as breaking change.
> 
> ```
> Renamed properties in bindings (model to chatModel and embeddingsModel, connectionName to searchConnectionName and storeConnectionName) and renamed ChatMessage to AssistantMessage. Updated managed identity support via the aiConnectionName property for Azure OpenAI
> ```

I wanted to create a base class, AssistantBaseAttribute, to serve as the foundation for AssistantAttribute, SemanticSearchAttribute, and TextCompletionAttribute.
 
One of the child classes uses two models - ChatModel and EmbeddingsModel, so I aimed to keep them distinct.
 
Additionally, a new property, AIConnectionName, was introduced for Azure OpenAI connection settings. Therefore, I changed the connection name property for EmbeddingStoreAttribute to storeConnectionName to clearly indicate it is for the embedding store. Similarly, I renamed another connectionName to searchConnectionName to specify it is for the search resource connection.

manvkaur commented on java-library/src/main/java/com/microsoft/azure/functions/openai/annotation/embeddings/EmbeddingsStoreOutput.java:
```diff
@@ -27,17 +27,21 @@
      * @return The model ID.
      */
-    String model() default "text-embedding-ada-002";
+    String embeddingsModel() default "text-embedding-ada-002";
```
Good catch, I have added ModelDefault constants for java-library.

manvkaur commented on src/WebJobs.Extensions.OpenAI/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,7 @@
+using System.Runtime.CompilerServices;
+
+// Make internals visible to the test project
+[assembly: InternalsVisibleTo("WebJobsOpenAIUnitTests")]
```
resolving in the same PR.

manvkaur commented on java-library/src/main/java/com/microsoft/azure/functions/openai/annotation/search/SemanticSearch.java:
```diff
@@ -69,10 +77,10 @@
      */
     String chatModel() default "gpt-3.5-turbo";
```
Are there any better models which should be default? 
Changing embedding models had impact in the past, it changes the embeddings generated.

manvkaur commented on src/Functions.Worker.Extensions.OpenAI/Embeddings/EmbeddingsContext.cs:
```diff
@@ -1,34 +1,30 @@
     /// Embeddings request sent to OpenAI.
     /// </summary>
-    public OpenAISDK.EmbeddingsOptions Request { get; set; }
+    public IList<string> Input { get; set; }
```
Valid point. Updated to Request/Response combination. I will talk in-person about the custom input context object.

TsuyoshiUshio commented on java-library/src/main/java/com/microsoft/azure/functions/openai/annotation/assistant/AssistantPost.java:
```diff
@@ -130,4 +130,11 @@
+     *
+     * @return {@code true} if the assistant is based on a reasoning model; {@code false} otherwise.
+     */
+    boolean isReasoningModel() default false;
```
Thank you for the update! My only concern is this design. This design currently works, however, essentially, this behavior depend on models. (e.g. o3, 4o ... etc.) If you want to switch, It might be on the model. I'd ping @cgillum for better design for this one (or ok for this boolean)

manvkaur replied:
Problem is reasoning models properties don't have parity with earlier models and I foresee SDK changing some behavior or configuration of these properties in future as par below comment in their ChangeLog
<img width="806" alt="image" src="https://github.com/user-attachments/assets/df2c4e64-7527-4e4c-bb4f-17bf0b392b8a" />
https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/openai/Azure.AI.OpenAI/CHANGELOG.md#features-added-1

cgillum replied:
I agree with Tsuyoshi's concern here. Customers may not understand when to configure this property - ideally it is automatic. Is there a middle ground where we can try to automatically detect, but customers can set a property to override the auto-detection if some future change to the models comes?

manvkaur replied:
Adding the offline discussion with Chris - Since users can choose a different deployment name for Azure OpenAI, we can't rely on hardcoded indexes of reasoning models to determine if a configured model is a reasoning model. I have added a README option to inform users about the `IsReasoningModel `property
